<center><h1>Analisis Exploratorio de los Datos - Parte 1 </h1></center>




En este apartado, procederemos a realizar la extracción de los datos. Luego de llevar a cabo un breve análisis exploratorio y descriptivo de estos, abordaremos temas como outliers, valores duplicados y valores faltantes. Además, nos ocuparemos de normalizar ciertas columnas para poder trabajar con ellas de manera óptima en la realización de gráficos pertinentes con sus respectivas variables. Se decidió dividir esta EDA en dos partes para una mayor practicidad y facilidad de lectura y organización. Esta primera abarcará lo mencionado anteriormente, y posteriormente, en el segundo documento [EDA_PARTE_2](EDA_Parte_2.ipynb), se llevarán a cabo los gráficos pertinentes.

# importaciones

In [67]:
import pandas as pd

%load_ext autoreload
%autoreload 2

from Utils import * 

import warnings
warnings.filterwarnings("ignore")

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


# ETL Datos Hechos exel Homicidios

In [107]:
homicidios_hechos = pd.read_excel('./Data/homicidios.xlsx', sheet_name='HECHOS')
homicidios_hechos.head()

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA HH  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00  4   
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00  1   
2  2016-0003           1 2016-01-03  2016   1   3  07:00:00  7   
3  2016-0004           1 2016-01-10  2016   1  10  00:00:00  0   
4  2016-0005           1 2016-01-21  2016   1  21  05:20:00  5   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE  ...  Altura  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...     NaN   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...     NaN   
2                         AV ENTRE RIOS 2034       AVENIDA  ...  2034.0   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...     NaN   
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...     NaN   

                                 Cruce  \
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                  DE LOS CORRALES AV.   
2                                  NaN   
3             VILLEGAS, CONRADO, GRAL.   
4              SAENZ PE?A, LUIS, PRES.   

                               Dirección Normalizada COMUNA  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                ENTRE RIOS AV. 2034      1   
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   

                                 XY (CABA)         pos x         pos y  \
0    Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1    Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2   Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3    Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4  Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   

    PARTICIPANTES VICTIMA    ACUSADO  
0       MOTO-AUTO    MOTO       AUTO  
1  AUTO-PASAJEROS    AUTO  PASAJEROS  
2       MOTO-AUTO    MOTO       AUTO  
3         MOTO-SD    MOTO         SD  
4  MOTO-PASAJEROS    MOTO  PASAJEROS  

[5 rows x 21 columns]

Datos generales:

In [108]:
print("columnas: ",homicidios_hechos.columns)

print("Cantidad de datos: ",homicidios_hechos.shape)

columnas:  Index(['ID', 'N_VICTIMAS', 'FECHA', 'AAAA', 'MM', 'DD', 'HORA', 'HH',
       'LUGAR_DEL_HECHO', 'TIPO_DE_CALLE', 'Calle', 'Altura', 'Cruce',
       'Dirección Normalizada', 'COMUNA', 'XY (CABA)', 'pos x', 'pos y',
       'PARTICIPANTES', 'VICTIMA', 'ACUSADO'],
      dtype='object')
Cantidad de datos:  (696, 21)


Estandarizo los nombres de las columnas

In [109]:
# Normalizar los nombres de las columnas
homicidios_hechos.columns = homicidios_hechos.columns.str.title()

# Cambiar los nombres específicos de columnas
homicidios_hechos = homicidios_hechos.rename(columns={'Aaaa': 'Ano', 'Mm': 'Mes', 'Dd': 'Dia'})

# Imprimir las columnas después de la normalización y cambio de nombres
print("Columnas después de la normalización y cambio de nombres:")
print(homicidios_hechos.columns)

Columnas después de la normalización y cambio de nombres:
Index(['Id', 'N_Victimas', 'Fecha', 'Ano', 'Mes', 'Dia', 'Hora', 'Hh',
       'Lugar_Del_Hecho', 'Tipo_De_Calle', 'Calle', 'Altura', 'Cruce',
       'Dirección Normalizada', 'Comuna', 'Xy (Caba)', 'Pos X', 'Pos Y',
       'Participantes', 'Victima', 'Acusado'],
      dtype='object')


# Datos Nulos
<a id='informacion'></a>

In [110]:
resumen_columnas(homicidios_hechos)

Columna  Cantidad de Datos  Cantidad de Nulos  \
0                      Id                696                  0   
1              N_Victimas                696                  0   
2                   Fecha                696                  0   
3                     Ano                696                  0   
4                     Mes                696                  0   
5                     Dia                696                  0   
6                    Hora                696                  0   
7                      Hh                696                  0   
8         Lugar_Del_Hecho                696                  0   
9           Tipo_De_Calle                696                  0   
10                  Calle                695                  1   
11                 Altura                129                567   
12                  Cruce                525                171   
13  Dirección Normalizada                688                  8   
14                 Comuna                696                  0   
15              Xy (Caba)                696                  0   
16                  Pos X                696                  0   
17                  Pos Y                696                  0   
18          Participantes                696                  0   
19                Victima                696                  0   
20                Acusado                696                  0   

    Porcentaje de Nulos  Cantidad de Valores Distintos   Tipo de Datos  
0              0.000000                            696          object  
1              0.000000                              3           int64  
2              0.000000                            598  datetime64[ns]  
3              0.000000                              6           int64  
4              0.000000                             12           int64  
5              0.000000                             31           int64  
6              0.000000                            375          object  
7              0.000000                             25          object  
8              0.000000                            683          object  
9              0.000000                              4          object  
10             0.143678                            279          object  
11            81.465517                            126         float64  
12            24.568966                            317          object  
13             1.149425                            635          object  
14             0.000000                             16           int64  
15             0.000000                            606          object  
16             0.000000                            605          object  
17             0.000000                            605          object  
18             0.000000                             41          object  
19             0.000000                             10          object  
20             0.000000                             10          object

como podemos observar, las unicas 3 columnas que poseen datos nulos son Cruce, Calle y Altura, esta ultima posee un 81% de datos nulos y es cuando el hecho se produce en algun punto de la calle es decir hace referencia a la altura de la calle, podemos decir con esto que es debido a que la mayoria de los hechos ocurren en las esquinas. Adicionalmente esta informacion se encuentra en las columnas "Lugar del hecho" y "Direccion normalizada". Se procede a eliminar esta columna y a ejemplificar

In [111]:
homicidios_hechos[['Altura', 'Lugar_Del_Hecho', 'Dirección Normalizada']][:3]

Altura                            Lugar_Del_Hecho  \
0     NaN  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ   
1     NaN          AV GRAL PAZ  Y AV DE LOS CORRALES   
2  2034.0                         AV ENTRE RIOS 2034   

                               Dirección Normalizada  
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...  
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.  
2                                ENTRE RIOS AV. 2034

In [112]:
# Se elimina la columna
homicidios_hechos = homicidios_hechos.drop('Altura', axis=1)
homicidios_hechos.columns

Index(['Id', 'N_Victimas', 'Fecha', 'Ano', 'Mes', 'Dia', 'Hora', 'Hh',
       'Lugar_Del_Hecho', 'Tipo_De_Calle', 'Calle', 'Cruce',
       'Dirección Normalizada', 'Comuna', 'Xy (Caba)', 'Pos X', 'Pos Y',
       'Participantes', 'Victima', 'Acusado'],
      dtype='object')

Columna **Cruce** es una variable categorica, de tipo string, que contiene 317 datos distintos y ademas un 24,6% de valores nulos. Esta variable contien los datos de cuando un hecho ocurre en un cruce, debido a esto se decide realizar una columna binaria, con un Si o No, dependiendo si el hecho ocurrio en un cruce o no y ademas mantener la columna actual para su posterios analisis y rellenar los NAN con 'No en cruce'.

In [113]:
# Se ven los valores nulos de cruce en relación con 'Lugar del hecho' y 'Tipo de calle'
homicidios_hechos[homicidios_hechos['Cruce'].isnull()][['Cruce', 'Lugar_Del_Hecho', 'Tipo_De_Calle']][:5]

Cruce                                    Lugar_Del_Hecho Tipo_De_Calle
2    NaN                                 AV ENTRE RIOS 2034       AVENIDA
9    NaN                                 AV ENTRE RIOS 1366       AVENIDA
14   NaN                                       SUIPACHA 156         CALLE
33   NaN                                          LIMA 1483         CALLE
35   NaN  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI KM....     AUTOPISTA

In [114]:
# Se ven los valores NO nulos de cruce en relación con 'Lugar del hecho' y 'Tipo de calle'
homicidios_hechos[homicidios_hechos['Cruce'].notnull()][['Cruce', 'Lugar_Del_Hecho', 'Tipo_De_Calle']][:5]

Cruce  \
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                  DE LOS CORRALES AV.   
3             VILLEGAS, CONRADO, GRAL.   
4              SAENZ PE?A, LUIS, PRES.   
5                         ESCALADA AV.   

                             Lugar_Del_Hecho Tipo_De_Calle  
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  
5             AV 27 DE FEBRERO Y AV ESCALADA       AVENIDA

In [115]:
# Rellenamos los NaN con el valor 'No en cruce'
homicidios_hechos['Cruce'].fillna('No en cruce', inplace=True)

# Cambiamos el nombre de la columna 'Cruce' por 'Nombre_Cruce'
homicidios_hechos.rename(columns={'Cruce': 'Nombre_Cruce'}, inplace=True)

# Creamos la columna 'Cruce' binaria donde contendrá valores 'Sí' o 'No'
homicidios_hechos['Cruce'] = 'Sí'  # Establecemos inicialmente todos los valores como 'Sí'
homicidios_hechos.loc[homicidios_hechos['Nombre_Cruce'] == 'No en cruce', 'Cruce'] = 'No'


In [116]:
# Se revia la columna
homicidios_hechos[['Cruce','Nombre_Cruce', 'Lugar_Del_Hecho', 'Tipo_De_Calle']]

Cruce                         Nombre_Cruce  \
0      Sí  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1      Sí                  DE LOS CORRALES AV.   
2      No                          No en cruce   
3      Sí             VILLEGAS, CONRADO, GRAL.   
4      Sí              SAENZ PE?A, LUIS, PRES.   
..    ...                                  ...   
691    Sí                                  MOM   
692    Sí                          LACARRA AV.   
693    Sí                              TERRADA   
694    No                          No en cruce   
695    No                          No en cruce   

                               Lugar_Del_Hecho Tipo_De_Calle  
0    AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  
1            AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  
2                           AV ENTRE RIOS 2034       AVENIDA  
3        AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  
4    AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  
..                                         ...           ...  
691                          AV. RIESTRA Y MOM       AVENIDA  
692                    AU DELLEPIANE Y LACARRA     AUTOPISTA  
693                        AV. GAONA Y TERRADA       AVENIDA  
694                         AV. EVA PERON 4071       AVENIDA  
695                    PADRE CARLOS MUJICA 709         CALLE  

[696 rows x 4 columns]

La tercer columna con valores faltantes es 'Dirección_Normalizada' tiene solo 8 datos nulos, lo que representa el 1.1% de los datos de la columna. Se orservan esos registros a continuación.

In [117]:
homicidios_hechos[homicidios_hechos['Dirección Normalizada'].isnull()][['Dirección Normalizada', 'Lugar_Del_Hecho', 'Tipo_De_Calle']][:10]

Dirección Normalizada                          Lugar_Del_Hecho  \
38                    NaN               AUTOPISTA LUGONES PK 10000   
106                   NaN         AU BUENOS AIRES - LA PLATA KM. 4   
119                   NaN                                       SD   
180                   NaN  AU PERITO MORENO Y RAMAL ENLACE AU1/AU6   
181                   NaN                       AU DELLEPIANE 2400   
313                   NaN                 AUTOPISTA LUGONES KM 4.7   
546                   NaN             LUGONES, LEOPOLDO AV. KM 6,1   
621                   NaN         AU BUENOS AIRES LA PLATA  KM 4,5   

    Tipo_De_Calle  
38      AUTOPISTA  
106     AUTOPISTA  
119         CALLE  
180     AUTOPISTA  
181     AUTOPISTA  
313     AUTOPISTA  
546     AUTOPISTA  
621     AUTOPISTA

No tenemos informacion que nos permita determinar un dato concreto de estos faltantes, pero no es conveniente eliminas los registros. Por lo tanto, se procede a rellenar estos valores con SD, indicando que es Sin Dato

In [118]:
# Se rellenan nulos con SD
homicidios_hechos['Dirección Normalizada'].fillna("SD", inplace=True)
# Se verifican los datos imputados
homicidios_hechos[homicidios_hechos['Dirección Normalizada']== 'SD'][['Dirección Normalizada', 'Lugar_Del_Hecho', 'Tipo_De_Calle']]

Dirección Normalizada                          Lugar_Del_Hecho  \
38                     SD               AUTOPISTA LUGONES PK 10000   
106                    SD         AU BUENOS AIRES - LA PLATA KM. 4   
119                    SD                                       SD   
180                    SD  AU PERITO MORENO Y RAMAL ENLACE AU1/AU6   
181                    SD                       AU DELLEPIANE 2400   
313                    SD                 AUTOPISTA LUGONES KM 4.7   
546                    SD             LUGONES, LEOPOLDO AV. KM 6,1   
621                    SD         AU BUENOS AIRES LA PLATA  KM 4,5   

    Tipo_De_Calle  
38      AUTOPISTA  
106     AUTOPISTA  
119         CALLE  
180     AUTOPISTA  
181     AUTOPISTA  
313     AUTOPISTA  
546     AUTOPISTA  
621     AUTOPISTA

# Datos duplicados

No se encontraron filas duplicadas en el conjunto de datos. La variable **'Id'** cuenta con 696 valores únicos. 
Revisa [AQUI](#informacion) la informacion

# Tipos de variables

In [119]:
#observamos si hay homogeneidad en las columnas con sus tipos de variables 
tipos_de_variables(homicidios_hechos)

{'Id': [str],
 'N_Victimas': [int],
 'Fecha': [pandas._libs.tslibs.timestamps.Timestamp],
 'Ano': [int],
 'Mes': [int],
 'Dia': [int],
 'Hora': [datetime.time, str, datetime.datetime],
 'Hh': [int, str],
 'Lugar_Del_Hecho': [str],
 'Tipo_De_Calle': [str],
 'Calle': [str, float],
 'Nombre_Cruce': [str],
 'Dirección Normalizada': [str],
 'Comuna': [int],
 'Xy (Caba)': [str],
 'Pos X': [str],
 'Pos Y': [str],
 'Participantes': [str],
 'Victima': [str],
 'Acusado': [str],
 'Cruce': [str]}

como podemos observar 'Hora', 'Hh' y 'Calle' contienen varios tipos de datos en la misma columna, por lo que procedemos a revisarlas y corregirlas. 

Columna 'Hora'

In [120]:

# Cantidad de valores por tipo de dato en la columna 'Hora'
conteo_tipos = homicidios_hechos['Hora'].apply(type).value_counts()

#muestra de cada tipo de dato
for tipo, cantidad in conteo_tipos.items():
    muestra = homicidios_hechos.loc[homicidios_hechos['Hora'].apply(type) == tipo, 'Hora'].head()
    print(f'Tipo: {tipo}, Cantidad: {cantidad}\nMuestra:\n{muestra}\n')

Tipo: <class 'datetime.time'>, Cantidad: 608
Muestra:
0    04:00:00
1    01:15:00
2    07:00:00
3    00:00:00
4    05:20:00
Name: Hora, dtype: object

Tipo: <class 'str'>, Cantidad: 85
Muestra:
275    22:50:00
276    16:40:00
277    10:12:00
278    09:40:00
279    22:30:00
Name: Hora, dtype: object

Tipo: <class 'datetime.datetime'>, Cantidad: 3
Muestra:
439    1900-01-02 00:00:00
441    1900-01-14 07:12:00
496    1900-01-04 18:40:00
Name: Hora, dtype: object



Dado que la mayoría de los datos son del tipo datetime.time, se decide convertir todos a este tipo de dato.

In [121]:
# Se cambia el tipo de dato
homicidios_hechos['Hora'] = homicidios_hechos['Hora'].apply(lambda x: convertir_a_time(x))
# Se verifica la cantidad de valores por tipo de dato en la columna 'hora'
print('Tipos de datos:')
print(homicidios_hechos['Hora'].apply(type).value_counts())
print('Registro con NoneType:')
homicidios_hechos[homicidios_hechos['Hora'].isna()]

Tipos de datos:
Hora
<class 'datetime.time'>    695
<class 'NoneType'>           1
Name: count, dtype: int64
Registro con NoneType:


Id  N_Victimas      Fecha   Ano  Mes  Dia  Hora  Hh  \
518  2019-0103           1 2019-12-18  2019   12   18  None  SD   

             Lugar_Del_Hecho Tipo_De_Calle  ... Nombre_Cruce  \
518  PAZ, GRAL. AV. Y GRIVEO      GRAL PAZ  ...       GRIVEO   

       Dirección Normalizada Comuna                               Xy (Caba)  \
518  PAZ, GRAL. AV. y GRIVEO     11  Point (94643.11254058 103831.57115061)   

            Pos X         Pos Y Participantes Victima Acusado Cruce  
518  -58.52169422  -34.59471640     MOTO-MOTO    MOTO    MOTO    Sí  

[1 rows x 21 columns]

se observa que no se encuentras datos en Hora y Hh. Para rellenar este dato con la moda de la columna Hora

In [122]:
# calcula la hora mas común
hora_moda = homicidios_hechos['Hora'].mode().iloc[0]
print(f'La hora mas común es: {hora_moda}')
# Se reemplaza el valor None por la hora más común
homicidios_hechos['Hora'].fillna(hora_moda, inplace=True)
# se verifica el tipo de dato para la columna
homicidios_hechos['Hora'].apply(type).value_counts()

La hora mas común es: 09:00:00


Hora
<class 'datetime.time'>    696
Name: count, dtype: int64

### Columna 'Hh'
procedomos a hacer le mismo tratamiento con la columna 'Hh'

In [123]:
# Se verifica el tipo de dato
homicidios_hechos['Hh'].apply(type).value_counts()

Hh
<class 'int'>    695
<class 'str'>      1
Name: count, dtype: int64

In [124]:
# Se imputa la hora moda al dato faltante
homicidios_hechos['Hh'] = homicidios_hechos['Hh'].apply(lambda x: int(hora_moda.hour) if x == "SD" else x)
# Se verifica el tipo de dato
homicidios_hechos['Hh'].apply(type).value_counts()

Hh
<class 'int'>    696
Name: count, dtype: int64

cambiamos el nombre de la columna, para una mejor identificacion

In [125]:
homicidios_hechos.rename(columns={'Hh': 'Hora_Entera'}, inplace=True)

### columna 'Calle'

revisamos esta columna ya que cuenta con dos tipos de datos

In [126]:
# Se verifica el tipo de dato
print('Tipos de datos:')
print(homicidios_hechos['Calle'].apply(type).value_counts())
# Se observa el registro con tipo de dato float
print('Registro con tipo float:')
homicidios_hechos[homicidios_hechos['Calle'].apply(lambda x: isinstance(x, float))]

Tipos de datos:
Calle
<class 'str'>      695
<class 'float'>      1
Name: count, dtype: int64
Registro con tipo float:


Id  N_Victimas      Fecha   Ano  Mes  Dia      Hora  Hora_Entera  \
119  2016-0151           1 2016-11-18  2016   11   18  20:35:00           20   

    Lugar_Del_Hecho Tipo_De_Calle  ... Nombre_Cruce Dirección Normalizada  \
119              SD         CALLE  ...  No en cruce                    SD   

    Comuna    Xy (Caba) Pos X Pos Y Participantes Victima Acusado Cruce  
119      0  Point (. .)     .     .     PEATON-SD  PEATON      SD    No  

[1 rows x 21 columns]

Se observa que no se cuenta con el dato de la calle donde ocurrió el hecho. Por lo tanto, se decide imputar *'SD'* a este valor faltante.

In [127]:
# Se imputa SD al dato faltante
homicidios_hechos['Calle'].fillna('SD', inplace=True)
# Se verifica el registro completado
homicidios_hechos[homicidios_hechos['Id']=='2016-0151']

Id  N_Victimas      Fecha   Ano  Mes  Dia      Hora  Hora_Entera  \
119  2016-0151           1 2016-11-18  2016   11   18  20:35:00           20   

    Lugar_Del_Hecho Tipo_De_Calle  ... Nombre_Cruce Dirección Normalizada  \
119              SD         CALLE  ...  No en cruce                    SD   

    Comuna    Xy (Caba) Pos X Pos Y Participantes Victima Acusado Cruce  
119      0  Point (. .)     .     .     PEATON-SD  PEATON      SD    No  

[1 rows x 21 columns]

verificamos los tipos de datos de las variables una vez mas

In [128]:
tipos_de_variables(homicidios_hechos)

{'Id': [str],
 'N_Victimas': [int],
 'Fecha': [pandas._libs.tslibs.timestamps.Timestamp],
 'Ano': [int],
 'Mes': [int],
 'Dia': [int],
 'Hora': [datetime.time],
 'Hora_Entera': [int],
 'Lugar_Del_Hecho': [str],
 'Tipo_De_Calle': [str],
 'Calle': [str],
 'Nombre_Cruce': [str],
 'Dirección Normalizada': [str],
 'Comuna': [int],
 'Xy (Caba)': [str],
 'Pos X': [str],
 'Pos Y': [str],
 'Participantes': [str],
 'Victima': [str],
 'Acusado': [str],
 'Cruce': [str]}

Columnas 'Victima' y 'Acusado'

En este punto se revisan si las categorías informadas por el diccionario de datos es la correcta.

In [129]:
homicidios_hechos['Victima'].unique()

array(['MOTO', 'AUTO', 'PEATON', 'SD', 'CARGAS', 'BICICLETA', 'PASAJEROS',
       'MOVIL', 'OBJETO FIJO', 'PEATON_MOTO'], dtype=object)

In [130]:
homicidios_hechos['Acusado'].unique()

array(['AUTO', 'PASAJEROS', 'SD', 'OBJETO FIJO', 'CARGAS', 'MOTO',
       'MULTIPLE', 'OTRO', 'BICICLETA', 'TREN'], dtype=object)

La columna 'Acusado' coincide con el diccionario, la columna Victima contiene dos variables 'Objeto Fijo' y 'Peaton Moto' que no estan en el diccionario de datos, procedemos a observarlos:

In [131]:
homicidios_hechos[homicidios_hechos['Victima'].isin(['OBJETO FIJO', 'PEATON_MOTO'])]

Id  N_Victimas      Fecha   Ano  Mes  Dia      Hora  Hora_Entera  \
230  2017-0108           2 2017-09-02  2017    9    2  04:53:08            4   
583  2020-0063           2 2020-12-05  2020   12    5  07:10:00            7   

              Lugar_Del_Hecho Tipo_De_Calle  ... Nombre_Cruce  \
230   AV. GRAL. PAZ Y MACHAIN      GRAL PAZ  ...      MACHAIN   
583  NUEVA YORK Y ALTA GRACIA         CALLE  ...  ALTA GRACIA   

        Dirección Normalizada Comuna                               Xy (Caba)  \
230  PAZ, GRAL. AV. y MACHAIN     12  Point (97098.48468623 109019.96106626)   
583  NUEVA YORK y ALTA GRACIA     11  Point (94080.62190808 102083.62453795)   

            Pos X         Pos Y     Participantes      Victima Acusado Cruce  
230  -58.49491054  -34.54795581  AUTO-OBJETO FIJO  OBJETO FIJO    AUTO    Sí  
583  -58.52783814  -34.61047001  PEATON_MOTO-MOTO  PEATON_MOTO    MOTO    Sí  

[2 rows x 21 columns]

procemos a imputar ambos registros como 'SD' debido a que no tenemos informacion para indicar el vehiculo de la victima. 

In [132]:
# Se cambia por OTRO
homicidios_hechos['Victima'] = homicidios_hechos['Victima'].replace({'OBJETO FIJO':'SD', 'PEATON_MOTO':'SD'})
# Se verifican esos cambios
homicidios_hechos[homicidios_hechos['Id'].isin(['2017-0108', '2020-0063'])]

Id  N_Victimas      Fecha   Ano  Mes  Dia      Hora  Hora_Entera  \
230  2017-0108           2 2017-09-02  2017    9    2  04:53:08            4   
583  2020-0063           2 2020-12-05  2020   12    5  07:10:00            7   

              Lugar_Del_Hecho Tipo_De_Calle  ... Nombre_Cruce  \
230   AV. GRAL. PAZ Y MACHAIN      GRAL PAZ  ...      MACHAIN   
583  NUEVA YORK Y ALTA GRACIA         CALLE  ...  ALTA GRACIA   

        Dirección Normalizada Comuna                               Xy (Caba)  \
230  PAZ, GRAL. AV. y MACHAIN     12  Point (97098.48468623 109019.96106626)   
583  NUEVA YORK y ALTA GRACIA     11  Point (94080.62190808 102083.62453795)   

            Pos X         Pos Y     Participantes Victima Acusado Cruce  
230  -58.49491054  -34.54795581  AUTO-OBJETO FIJO      SD    AUTO    Sí  
583  -58.52783814  -34.61047001  PEATON_MOTO-MOTO      SD    MOTO    Sí  

[2 rows x 21 columns]

### Variables geograficas

Durante el [análisis anterior](#informacion), se pudo ver que hay faltantes de información geográfica en las columnas **'XY_(CABA)'**, **'Pos_x'** y **'Pos_y'**. Se observan esos registros.

In [133]:
homicidios_hechos[homicidios_hechos['Xy (Caba)']=='Point (. .)']

Id  N_Victimas      Fecha   Ano  Mes  Dia      Hora  Hora_Entera  \
35   2016-0049           1 2016-04-17  2016    4   17  00:00:00            0   
38   2016-0052           1 2016-04-20  2016    4   20  20:00:00           20   
71   2016-0096           1 2016-07-25  2016    7   25  07:00:00            7   
106  2016-0136           1 2016-10-25  2016   10   25  00:00:00            0   
119  2016-0151           1 2016-11-18  2016   11   18  20:35:00           20   
139  2016-0174           1 2016-12-27  2016   12   27  00:00:00            0   
176  2017-0042           1 2017-04-10  2017    4   10  09:00:00            9   
180  2017-0050           2 2017-04-28  2017    4   28  11:08:08           11   
181  2017-0051           1 2017-05-01  2017    5    1  03:47:47            3   
256  2017-0140           1 2017-11-19  2017   11   19  23:22:17           23   
313  2018-0039           1 2018-04-21  2018    4   21  22:15:00           22   
546  2020-0026           1 2020-05-17  2020    5   17  06:40:00            6   
559  2020-0039           1 2020-09-01  2020    9    1  19:17:42           19   
621  2021-0023           1 2021-03-01  2021    3    1  09:20:00            9   

                                       Lugar_Del_Hecho Tipo_De_Calle  ...  \
35   AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI KM....     AUTOPISTA  ...   
38                          AUTOPISTA LUGONES PK 10000     AUTOPISTA  ...   
71        AUTOPISTA DELLEPIANE LUIS TTE. GRAL. KM. 2,3     AUTOPISTA  ...   
106                   AU BUENOS AIRES - LA PLATA KM. 4     AUTOPISTA  ...   
119                                                 SD         CALLE  ...   
139                               AUTOPISTA 25 DE MAYO     AUTOPISTA  ...   
176                      AV. LEOPOLDO LUGONES PKM 6900      GRAL PAZ  ...   
180            AU PERITO MORENO Y RAMAL ENLACE AU1/AU6     AUTOPISTA  ...   
181                                 AU DELLEPIANE 2400     AUTOPISTA  ...   
256                        AU ARTURO FRONDIZI PKM 3100     AUTOPISTA  ...   
313                           AUTOPISTA LUGONES KM 4.7     AUTOPISTA  ...   
546                       LUGONES, LEOPOLDO AV. KM 6,1     AUTOPISTA  ...   
559                                    MURGUIONDO 2700         CALLE  ...   
621                   AU BUENOS AIRES LA PLATA  KM 4,5     AUTOPISTA  ...   

    Nombre_Cruce                       Dirección Normalizada Comuna  \
35   No en cruce  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI      4   
38   No en cruce                                          SD     13   
71   No en cruce        AUTOPISTA DELLEPIANE LUIS TTE. GRAL.      8   
106  No en cruce                                          SD      4   
119  No en cruce                                          SD      0   
139  No en cruce                        AUTOPISTA 25 DE MAYO      0   
176  No en cruce                       LUGONES, LEOPOLDO AV.     14   
180  No en cruce                                          SD      9   
181  No en cruce                                          SD      7   
256  No en cruce  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI      4   
313  No en cruce                                          SD     14   
546  No en cruce                                          SD     14   
559  No en cruce                                  MURGUIONDO      9   
621  No en cruce                                          SD      4   

       Xy (Caba)               Pos X               Pos Y     Participantes  \
35   Point (. .)  -58.37714647568196  -34.63657525428238             SD-SD   
38   Point (. .)                   .                   .           MOTO-SD   
71   Point (. .)  -58.47433193007387  -34.66684950051973       MOTO-CARGAS   
106  Point (. .)                   .                   .       MOTO-CARGAS   
119  Point (. .)                   .                   .         PEATON-SD   
139  Point (. .)                   .                   .             SD-SD   
176  Point (. .)                   .       

 Se encontraron 14 registros de los cuales no peseemos informacion para completarlos, procedemos a ponerle un valor 0 para proximamente poder excluirlo en el dashboard

In [134]:
homicidios_hechos['Pos X'] = homicidios_hechos['Pos X'].replace('.', 0)
homicidios_hechos['Pos Y'] = homicidios_hechos['Pos Y'].replace('.', 0)
homicidios_hechos['Xy (Caba)'] = homicidios_hechos['Xy (Caba)'].replace('Point (. .)', 0)

In [135]:
#Verificamos con una fila
homicidios_hechos[homicidios_hechos['Id']=='2016-0052']

Id  N_Victimas      Fecha   Ano  Mes  Dia      Hora  Hora_Entera  \
38  2016-0052           1 2016-04-20  2016    4   20  20:00:00           20   

               Lugar_Del_Hecho Tipo_De_Calle  ... Nombre_Cruce  \
38  AUTOPISTA LUGONES PK 10000     AUTOPISTA  ...  No en cruce   

   Dirección Normalizada Comuna  Xy (Caba) Pos X Pos Y Participantes Victima  \
38                    SD     13          0     0     0       MOTO-SD    MOTO   

   Acusado Cruce  
38      SD    No  

[1 rows x 21 columns]

procedeemos a revisar nuestras variables nuevamente para chequear que todas contengan los 696 datos y sin nulos

In [136]:
resumen_columnas(homicidios_hechos)

Columna  Cantidad de Datos  Cantidad de Nulos  \
0                      Id                696                  0   
1              N_Victimas                696                  0   
2                   Fecha                696                  0   
3                     Ano                696                  0   
4                     Mes                696                  0   
5                     Dia                696                  0   
6                    Hora                696                  0   
7             Hora_Entera                696                  0   
8         Lugar_Del_Hecho                696                  0   
9           Tipo_De_Calle                696                  0   
10                  Calle                696                  0   
11           Nombre_Cruce                696                  0   
12  Dirección Normalizada                696                  0   
13                 Comuna                696                  0   
14              Xy (Caba)                696                  0   
15                  Pos X                696                  0   
16                  Pos Y                696                  0   
17          Participantes                696                  0   
18                Victima                696                  0   
19                Acusado                696                  0   
20                  Cruce                696                  0   

    Porcentaje de Nulos  Cantidad de Valores Distintos   Tipo de Datos  
0                   0.0                            696          object  
1                   0.0                              3           int64  
2                   0.0                            598  datetime64[ns]  
3                   0.0                              6           int64  
4                   0.0                             12           int64  
5                   0.0                             31           int64  
6                   0.0                            322          object  
7                   0.0                             24           int64  
8                   0.0                            683          object  
9                   0.0                              4          object  
10                  0.0                            280          object  
11                  0.0                            318          object  
12                  0.0                            636          object  
13                  0.0                             16           int64  
14                  0.0                            606          object  
15                  0.0                            605          object  
16                  0.0                            605          object  
17                  0.0                             41          object  
18                  0.0                              8          object  
19                  0.0                             10          object  
20                  0.0                              2          object

# Victimas excel homicidios

In [137]:
homicidios_victimas = pd.read_excel('./Data/homicidios.xlsx', sheet_name='VICTIMAS')
homicidios_victimas.head()

ID_hecho      FECHA  AAAA  MM  DD        ROL VICTIMA       SEXO EDAD  \
0  2016-0001 2016-01-01  2016   1   1  CONDUCTOR    MOTO  MASCULINO   19   
1  2016-0002 2016-01-02  2016   1   2  CONDUCTOR    AUTO  MASCULINO   70   
2  2016-0003 2016-01-03  2016   1   3  CONDUCTOR    MOTO  MASCULINO   30   
3  2016-0004 2016-01-10  2016   1  10  CONDUCTOR    MOTO  MASCULINO   18   
4  2016-0005 2016-01-21  2016   1  21  CONDUCTOR    MOTO  MASCULINO   29   

   FECHA_FALLECIMIENTO  
0  2016-01-01 00:00:00  
1  2016-01-02 00:00:00  
2  2016-01-03 00:00:00  
3                   SD  
4  2016-02-01 00:00:00

procedemos a ver un resumen de los datos

In [138]:
# Se observan las columnas de dataset
homicidios_victimas.columns

Index(['ID_hecho', 'FECHA', 'AAAA', 'MM', 'DD', 'ROL', 'VICTIMA', 'SEXO',
       'EDAD', 'FECHA_FALLECIMIENTO'],
      dtype='object')

In [139]:
resumen_columnas(homicidios_victimas)

Columna  Cantidad de Datos  Cantidad de Nulos  \
0             ID_hecho                717                  0   
1                FECHA                717                  0   
2                 AAAA                717                  0   
3                   MM                717                  0   
4                   DD                717                  0   
5                  ROL                717                  0   
6              VICTIMA                717                  0   
7                 SEXO                717                  0   
8                 EDAD                717                  0   
9  FECHA_FALLECIMIENTO                717                  0   

   Porcentaje de Nulos  Cantidad de Valores Distintos   Tipo de Datos  
0                  0.0                            696          object  
1                  0.0                            598  datetime64[ns]  
2                  0.0                              6           int64  
3                  0.0                             12           int64  
4                  0.0                             31           int64  
5                  0.0                              5          object  
6                  0.0                              8          object  
7                  0.0                              3          object  
8                  0.0                             86          object  
9                  0.0                            563          object

cambio de nombre de las columnas Año, Mes y Dia, y estandarizacion de nombres de columnas

In [140]:
# Se coloca la primera en mayúscula
homicidios_victimas.columns = [x.capitalize() for x in homicidios_victimas.columns]
# Se reemplazan los guiones por espacios
homicidios_victimas.columns = homicidios_victimas.columns.str.replace('_', ' ')
# Se renombran algunas columnas
homicidios_victimas = homicidios_victimas.rename(columns={'Id hecho': 'Id',
                                                      'Aaaa':'Año',
                                                      'Mm':'Mes',
                                                      'Dd':'Día',
                                                      'Victima':'Víctima'})
homicidios_victimas.columns


Index(['Id', 'Fecha', 'Año', 'Mes', 'Día', 'Rol', 'Víctima', 'Sexo', 'Edad',
       'Fecha fallecimiento'],
      dtype='object')

### Datos Nulos

Desde el reporte se informa que no se encuentran valores nulos. A continuación, se verifica esto.

In [141]:
# Verifica si hay valores nulos en homicidios_victimas
homicidios_victimas.isna().any().any()

False

### Datos Duplicados

se verifica que no hay datos duplicados

In [142]:
# Se buscan las filas repetidas
homicidios_victimas[homicidios_victimas.duplicated()]

Empty DataFrame
Columns: [Id, Fecha, Año, Mes, Día, Rol, Víctima, Sexo, Edad, Fecha fallecimiento]
Index: []

Columna 'Id' como podemos observar esta columna contiene datos duplicados pero son debido a que en el mismo hecho con varias victimas, debido a esto se conservan todas las filas

In [143]:
# Encuentra los valores duplicados en la columna especificada
duplicados = homicidios_victimas[homicidios_victimas.duplicated(subset='Id', keep=False)]

# Muestra los resultados
print(duplicados)

            Id      Fecha   Año  Mes  Día                   Rol Víctima  \
29   2016-0041 2016-03-29  2016    3   29             CONDUCTOR    MOTO   
30   2016-0041 2016-03-29  2016    3   29  PASAJERO_ACOMPAÑANTE    MOTO   
98   2016-0126 2016-09-18  2016    9   18             CONDUCTOR    AUTO   
99   2016-0126 2016-09-18  2016    9   18  PASAJERO_ACOMPAÑANTE    AUTO   
163  2017-0026 2017-02-26  2017    2   26  PASAJERO_ACOMPAÑANTE    AUTO   
164  2017-0026 2017-02-26  2017    2   26             CONDUCTOR    AUTO   
173  2017-0035 2017-03-23  2017    3   23             CONDUCTOR    AUTO   
174  2017-0035 2017-03-23  2017    3   23  PASAJERO_ACOMPAÑANTE    AUTO   
175  2017-0035 2017-03-23  2017    3   23  PASAJERO_ACOMPAÑANTE    AUTO   
176  2017-0036 2017-03-29  2017    3   29  PASAJERO_ACOMPAÑANTE    MOTO   
177  2017-0036 2017-03-29  2017    3   29             CONDUCTOR    MOTO   
186  2017-0050 2017-04-28  2017    4   28             CONDUCTOR    MOTO   
187  2017-0050 2017-04-28

### Tipos de variables

Observamos que las columnas 'Edad' y 'Fecha Fallecimiento', presentan una irregularidad en sus tipos de variables, se procede a analizarlas y asignar un unico tipo

In [144]:
tipos_de_variables(homicidios_victimas)

{'Id': [str],
 'Fecha': [pandas._libs.tslibs.timestamps.Timestamp],
 'Año': [int],
 'Mes': [int],
 'Día': [int],
 'Rol': [str],
 'Víctima': [str],
 'Sexo': [str],
 'Edad': [int, str],
 'Fecha fallecimiento': [datetime.datetime, str]}

### Columna 'Edad'

In [145]:
# Cantidad de valores por tipo de dato en la columna 'edad'
tipos_datos = homicidios_victimas['Edad'].apply(type).value_counts()
print('Los tipos de datos son:')
print(tipos_datos)
print(f'Los datos {tipos_datos.index[1]} representan el {round((tipos_datos[1]/tipos_datos.sum()),2)}')

Los tipos de datos son:
Edad
<class 'int'>    664
<class 'str'>     53
Name: count, dtype: int64
Los datos <class 'str'> representan el 0.07


se procede a observar los datos de tipo string

In [146]:
homicidios_victimas[homicidios_victimas['Edad'].apply(lambda x: isinstance(x, str))]

Id      Fecha   Año  Mes  Día                   Rol    Víctima  \
30   2016-0041 2016-03-29  2016    3   29  PASAJERO_ACOMPAÑANTE       MOTO   
33   2016-0045 2016-04-11  2016    4   11             CONDUCTOR       MOTO   
35   2016-0048 2016-04-15  2016    4   15                PEATON     PEATON   
36   2016-0049 2016-04-17  2016    4   17                    SD         SD   
39   2016-0052 2016-04-20  2016    4   20                    SD       MOTO   
55   2016-0077 2016-06-13  2016    6   13                PEATON     PEATON   
63   2016-0085 2016-06-29  2016    6   29                    SD       MOTO   
72   2016-0096 2016-07-25  2016    7   25             CONDUCTOR       MOTO   
89   2016-0115 2016-09-02  2016    9    2                    SD         SD   
93   2016-0119 2016-09-04  2016    9    4  PASAJERO_ACOMPAÑANTE         SD   
108  2016-0136 2016-10-25  2016   10   25             CONDUCTOR       MOTO   
119  2016-0148 2016-11-16  2016   11   16             CONDUCTOR       MOTO   
121  2016-0151 2016-11-18  2016   11   18                PEATON     PEATON   
138  2016-0171 2016-12-25  2016   12   25             CONDUCTOR       MOTO   
141  2016-0174 2016-12-27  2016   12   27                    SD         SD   
142  2016-0175 2016-12-27  2016   12   27                PEATON     PEATON   
165  2017-0027 2017-02-26  2017    2   26                PEATON     PEATON   
176  2017-0036 2017-03-29  2017    3   29  PASAJERO_ACOMPAÑANTE       MOTO   
182  2017-0042 2017-04-10  2017    4   10             CONDUCTOR       MOTO   
225  2017-0093 2017-07-25  2017    7   25                PEATON     PEATON   
243  2017-0112 2017-09-10  2017    9   10  PASAJERO_ACOMPAÑANTE       AUTO   
275  2017-0151 2017-12-04  2017   12    4                PEATON     PEATON   
277  2017-0153 2017-12-11  2017   12   11  PASAJERO_ACOMPAÑANTE       MOTO   
299  2018-0014 2018-01-31  2018    1   31                PEATON     PEATON   
328  2018-0040 2018-04-27  2018    4   27                PEATON     PEATON   
359  2018-0071 2018-07-10  2018    7   10             CONDUCTOR       MOTO   
367  2018-0079 2018-07-23  2018    7   23                PEATON     PEATON   
371  2018-0083 2018-08-02  2018    8    2             CONDUCTOR       MOTO   
377  2018-0088 2018-08-17  2018    8   17             CONDUCTOR       MOTO   
391  2018-0101 2018-09-21  2018    9   21             CONDUCTOR       AUTO   
421  2018-0130 2018-12-03  2018   12    3             CONDUCTOR       MOTO   
444  2019-0010 2019-02-02  2019    2    2                PEATON     PEATON   
446  2019-0011 2019-02-10  2019    2   10                PEATON     PEATON   
454  2019-0019 2019-03-06  2019    3    6             CONDUCTOR       MOTO   
457  2019-0022 2019-03-15  2019    3   15  PASAJERO_ACOMPAÑANTE       AUTO   
461  2019-0026 2019-03-28  2019    3   28                PEATON     PEATON   
463  2019-0028 2019-04-06  2019    4    6                PEATON     PEATON   
469  2019-0034 2019-04-21  2019    4   21             CONDUCTOR       MOTO   
479  2019-0043 2019-05-21  2019    5   21              CICLISTA  BICICLETA   
498  2019-0063 2019-08-04  2019    8    4  PASAJERO_ACOMPAÑANTE       MOTO   
506  2019-0071 2019-08-27  2019    8   27                PEATON     PEATON   
529  2019-0094 2019-11-18  2019   11   18  PASAJERO_ACOMPAÑANTE       MOTO   
530  2019-0095 2019-11-20  2019   11   20             CONDUCTOR       MOTO   
548  2020-0010 2020-02-08  2020    2    8             CONDUCTOR       MOTO   
552  2020-0014 2020-02-27  2020    2   27             CONDUCTOR       MOTO   
555  2020-0017 2020-03-05  2020    3    5                PEATON     PEATON   
563  2020-0025 2020-05-12  2020    5   12                PEATON     PEATON   
572  2020-0034 2020-08-02  2020    8    2                PEATON     PEATON   
579  2020-0040 2020-09-12  2020    9   12             CONDUCTOR       AUTO   
631  2021-0012 2021-01-29  2021    1   29                PEATON     PEATON   
639  2021-0020 2021-02-21  2021   

Como se observan faltantes tanto en la columna 'Sexo' y 'Edad' se completaran esta ultima con el promedio teniendo cuenta el Sexo de la victima, debido a esto se procede a rellenar la columna Sexo

In [147]:
imputa_valor_mas_frecuente(homicidios_victimas,'Sexo')

Muestra antes de la imputación:
            Id      Fecha   Año  Mes  Día        Rol Víctima Sexo Edad  \
36   2016-0049 2016-04-17  2016    4   17         SD      SD   SD   SD   
39   2016-0052 2016-04-20  2016    4   20         SD    MOTO   SD   SD   
108  2016-0136 2016-10-25  2016   10   25  CONDUCTOR    MOTO   SD   SD   
121  2016-0151 2016-11-18  2016   11   18     PEATON  PEATON   SD   SD   
138  2016-0171 2016-12-25  2016   12   25  CONDUCTOR    MOTO   SD   SD   

    Fecha fallecimiento  
36                   SD  
39                   SD  
108                  SD  
121                  SD  
138                  SD  

Muestra después de la imputación:
          Id      Fecha   Año  Mes  Día        Rol Víctima       Sexo Edad  \
0  2016-0001 2016-01-01  2016    1    1  CONDUCTOR    MOTO  MASCULINO   19   
1  2016-0002 2016-01-02  2016    1    2  CONDUCTOR    AUTO  MASCULINO   70   
2  2016-0003 2016-01-03  2016    1    3  CONDUCTOR    MOTO  MASCULINO   30   
3  2016-0004 2016-01

Ahora imputamos la edad, de acuerdo al Sexo

In [148]:
imputar_edad_promedio_por_sexo(homicidios_victimas)

La edad promedio de Femenino es 51 y de Masculino es 40


revisamos lo hecho en las columnas

In [149]:
id_hecho_ejemplos = ['2016-0045', '2016-0048', '2016-0049']
homicidios_victimas[homicidios_victimas['Id'].isin(id_hecho_ejemplos)]

Id      Fecha   Año  Mes  Día        Rol Víctima       Sexo  Edad  \
33  2016-0045 2016-04-11  2016    4   11  CONDUCTOR    MOTO  MASCULINO    39   
35  2016-0048 2016-04-15  2016    4   15     PEATON  PEATON   FEMENINO    50   
36  2016-0049 2016-04-17  2016    4   17         SD      SD  MASCULINO    39   

   Fecha fallecimiento  
33                  SD  
35                  SD  
36                  SD

Vemos que 'Rol' y 'Victima' tambien tienen faltantes de datos. Se revisan la cantidad de SD de cada una

In [150]:
print(f"La cantidad de SD en 'rol' es de {len(homicidios_victimas[homicidios_victimas['Rol']=='SD'])}")
print(f"La cantidad de SD en 'victima' es de {len(homicidios_victimas[homicidios_victimas['Víctima']=='SD'])}")

La cantidad de SD en 'rol' es de 11
La cantidad de SD en 'victima' es de 9


Se comprueba que la cantidad de datos donde falta informacion es poca, se procede a rellenar con el valor mas frecuente para cada columna

In [151]:
# Imputa valor mas frecuente en la columna rol
imputa_valor_mas_frecuente(homicidios_victimas, 'Rol')

Muestra antes de la imputación:
           Id      Fecha   Año  Mes  Día Rol Víctima       Sexo  Edad  \
36  2016-0049 2016-04-17  2016    4   17  SD      SD  MASCULINO    39   
39  2016-0052 2016-04-20  2016    4   20  SD    MOTO  MASCULINO    39   
63  2016-0085 2016-06-29  2016    6   29  SD    MOTO  MASCULINO    39   
77  2016-0101 2016-08-07  2016    8    7  SD      SD  MASCULINO    67   
89  2016-0115 2016-09-02  2016    9    2  SD      SD  MASCULINO    39   

   Fecha fallecimiento  
36                  SD  
39                  SD  
63                  SD  
77                  SD  
89                  SD  

Muestra después de la imputación:
          Id      Fecha   Año  Mes  Día        Rol Víctima       Sexo  Edad  \
0  2016-0001 2016-01-01  2016    1    1  CONDUCTOR    MOTO  MASCULINO    19   
1  2016-0002 2016-01-02  2016    1    2  CONDUCTOR    AUTO  MASCULINO    70   
2  2016-0003 2016-01-03  2016    1    3  CONDUCTOR    MOTO  MASCULINO    30   
3  2016-0004 2016-01-10  201

In [152]:
imputa_valor_mas_frecuente(homicidios_victimas,'Víctima')

Muestra antes de la imputación:
            Id      Fecha   Año  Mes  Día                   Rol Víctima  \
36   2016-0049 2016-04-17  2016    4   17             CONDUCTOR      SD   
77   2016-0101 2016-08-07  2016    8    7             CONDUCTOR      SD   
89   2016-0115 2016-09-02  2016    9    2             CONDUCTOR      SD   
93   2016-0119 2016-09-04  2016    9    4  PASAJERO_ACOMPAÑANTE      SD   
141  2016-0174 2016-12-27  2016   12   27             CONDUCTOR      SD   

          Sexo  Edad Fecha fallecimiento  
36   MASCULINO    39                  SD  
77   MASCULINO    67                  SD  
89   MASCULINO    39                  SD  
93    FEMENINO    50                  SD  
141  MASCULINO    39                  SD  

Muestra después de la imputación:
          Id      Fecha   Año  Mes  Día        Rol Víctima       Sexo  Edad  \
0  2016-0001 2016-01-01  2016    1    1  CONDUCTOR    MOTO  MASCULINO    19   
2  2016-0003 2016-01-03  2016    1    3  CONDUCTOR    MOTO  MASCUL

Revisamos nuevamente las filas anteriores

In [153]:
id_hecho_ejemplos = ['2016-0045', '2016-0048', '2016-0049']
homicidios_victimas[homicidios_victimas['Id'].isin(id_hecho_ejemplos)]

Id      Fecha   Año  Mes  Día        Rol Víctima       Sexo  Edad  \
33  2016-0045 2016-04-11  2016    4   11  CONDUCTOR    MOTO  MASCULINO    39   
35  2016-0048 2016-04-15  2016    4   15     PEATON  PEATON   FEMENINO    50   
36  2016-0049 2016-04-17  2016    4   17  CONDUCTOR    MOTO  MASCULINO    39   

   Fecha fallecimiento  
33                  SD  
35                  SD  
36                  SD

Por Ultimo, la columna 'Fecha fallecimientos' posee dos tipos de datos y por lo que podemos observar en el ejemplo de arriba es debido a los datos faltantes. Como el objetivo de este proyecto es encontrar patrones que permitan tomar decisiones para disminuit los accidentes fatales, conocer esta fecha no nos aporta informacion al analisis, agregando como se observa abajo no hay un patron de repetision en cuanto a fechas de fallecimiento, procedemos a eliminarla.

In [154]:
top_10_valores_repetidos(homicidios_victimas,'Fecha fallecimiento')

Top 10 de valores más repetidos en la columna 'Fecha fallecimiento':
Fecha fallecimiento
SD                     68
2017-02-26 00:00:00     3
2017-01-16 00:00:00     3
2020-12-25 00:00:00     3
2019-12-18 00:00:00     3
2017-11-19 00:00:00     3
2018-04-27 00:00:00     3
2018-08-20 00:00:00     3
2018-08-03 00:00:00     3
2020-12-20 00:00:00     3
Name: count, dtype: int64


In [155]:
homicidios_victimas = homicidios_victimas.drop('Fecha fallecimiento', axis=1)
homicidios_victimas.head(3)

Id      Fecha   Año  Mes  Día        Rol Víctima       Sexo  Edad
0  2016-0001 2016-01-01  2016    1    1  CONDUCTOR    MOTO  MASCULINO    19
1  2016-0002 2016-01-02  2016    1    2  CONDUCTOR    AUTO  MASCULINO    70
2  2016-0003 2016-01-03  2016    1    3  CONDUCTOR    MOTO  MASCULINO    30

se verificas una ves mas los tipos de datos del conjunto de datos

In [156]:
tipos_de_variables(homicidios_victimas)

{'Id': [str],
 'Fecha': [pandas._libs.tslibs.timestamps.Timestamp],
 'Año': [int],
 'Mes': [int],
 'Día': [int],
 'Rol': [str],
 'Víctima': [str],
 'Sexo': [str],
 'Edad': [int]}

### Eliminacion de columnas

Dado que este conjunto de datos se unirá con el `homicidios_hechos`, se pueden eliminar las columnas 'Fecha', 'Año', 'Mes', 'Día' y 'Víctima', porque es la misma información que contiene `homicidios_hechos`. 

In [157]:
# Se eliminan las columnas repetidas
homicidios_victimas = homicidios_victimas.drop(['Fecha', 'Año', 'Mes', 'Día', 'Víctima'], axis=1)
homicidios_victimas.columns

Index(['Id', 'Rol', 'Sexo', 'Edad'], dtype='object')

# Unificacion de la Informacion

procedemos a unificar los dos conjuntos de datos limpiados anteriormente

In [158]:
df_homicidios = homicidios_victimas.merge(homicidios_hechos, on='Id', how='left')
df_homicidios.head(4)

Id        Rol       Sexo  Edad  N_Victimas      Fecha   Ano  Mes  \
0  2016-0001  CONDUCTOR  MASCULINO    19           1 2016-01-01  2016    1   
1  2016-0002  CONDUCTOR  MASCULINO    70           1 2016-01-02  2016    1   
2  2016-0003  CONDUCTOR  MASCULINO    30           1 2016-01-03  2016    1   
3  2016-0004  CONDUCTOR  MASCULINO    18           1 2016-01-10  2016    1   

   Dia      Hora  ...                         Nombre_Cruce  \
0    1  04:00:00  ...  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1    2  01:15:00  ...                  DE LOS CORRALES AV.   
2    3  07:00:00  ...                          No en cruce   
3   10  00:00:00  ...             VILLEGAS, CONRADO, GRAL.   

                               Dirección Normalizada Comuna  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                ENTRE RIOS AV. 2034      1   
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   

                                Xy (Caba)         Pos X         Pos Y  \
0   Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1   Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2  Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3   Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   

    Participantes Victima    Acusado Cruce  
0       MOTO-AUTO    MOTO       AUTO    Sí  
1  AUTO-PASAJEROS    AUTO  PASAJEROS    Sí  
2       MOTO-AUTO    MOTO       AUTO    No  
3         MOTO-SD    MOTO         SD    Sí  

[4 rows x 24 columns]

verificamos los datos

In [159]:
resumen_columnas(df_homicidios)

Columna  Cantidad de Datos  Cantidad de Nulos  \
0                      Id                717                  0   
1                     Rol                717                  0   
2                    Sexo                717                  0   
3                    Edad                717                  0   
4              N_Victimas                717                  0   
5                   Fecha                717                  0   
6                     Ano                717                  0   
7                     Mes                717                  0   
8                     Dia                717                  0   
9                    Hora                717                  0   
10            Hora_Entera                717                  0   
11        Lugar_Del_Hecho                717                  0   
12          Tipo_De_Calle                717                  0   
13                  Calle                717                  0   
14           Nombre_Cruce                717                  0   
15  Dirección Normalizada                717                  0   
16                 Comuna                717                  0   
17              Xy (Caba)                717                  0   
18                  Pos X                717                  0   
19                  Pos Y                717                  0   
20          Participantes                717                  0   
21                Victima                717                  0   
22                Acusado                717                  0   
23                  Cruce                717                  0   

    Porcentaje de Nulos  Cantidad de Valores Distintos   Tipo de Datos  
0                   0.0                            696          object  
1                   0.0                              4          object  
2                   0.0                              2          object  
3                   0.0                             85           int32  
4                   0.0                              3           int64  
5                   0.0                            598  datetime64[ns]  
6                   0.0                              6           int64  
7                   0.0                             12           int64  
8                   0.0                             31           int64  
9                   0.0                            322          object  
10                  0.0                             24           int64  
11                  0.0                            683          object  
12                  0.0                              4          object  
13                  0.0                            280          object  
14                  0.0                            318          object  
15                  0.0                            636          object  
16                  0.0                             16           int64  
17                  0.0                            606          object  
18                  0.0                            605          object  
19                  0.0                            605          object  
20                  0.0                             41          object  
21                  0.0                              8          object  
22                  0.0                             10          object  
23                  0.0                              2          object

se observa que esta correcto, se procede a guardar y posteriormente a trabajar con este conjunto de datos para su evalucion que permitan hacer recomendacion para disminuir la cantidad de accidentes fatales. Procedemos a cargar el dataframe

In [160]:
archivo_limpio = 'Data/homicidios_limpio.csv'
df_homicidios.to_csv(archivo_limpio, index=False, encoding='utf-8')